### ISSUE TITLE
Stateful wrappers

### ISSUE THREAD

**Issue by adamschoenemann** [testing-library/react-hooks-testing-library#809]
First of all thanks for providing this nice library.

I want to test some hooks that use context. So I supply a `wrapper` that renders the context providers. However, some of these providers also use `setState` inside `useEffect`. This means that I get lots of `An update to ... inside a test was not wrapped in act(...)` warnings, even though my tests pass fine.

I've looked through all the other issues to see if someone else has this problem, but I couldn't find anything.
I'm kinda stumped on how to proceed from here... Any suggestions?

Thanks!

**Comment by mpeyper** [comment]
The wrapper gets rendered with the test component that renders the hook, so any state changes in the wrapper should also get flushed with the same `act` calls being used to make changes to the hook’s state as well (or whatever it is exposing to update the wrapper). Unfortunately this means those updates are subject to the same `act` rules as the hook, for better or worse.

Can you share some code so I can understand the setup and scenario a bit better?

**Comment by adamschoenemann** [comment]
Thanks for the swift reply! I'll try and cook up a minimal example ASAP.

**Comment by adamschoenemann** [comment]
OK, so the behaviour I'm seeing is *very weird*.

Given this test:

```typescript
async function loadData() {}
const MyContext = React.createContext(undefined as unknown as any);

const MyProvider: React.FC = (props) => {
  const [state, setState] = React.useState({ foo: 0 });
  React.useEffect(() => {
    async function load() {
      await loadData();
      setState({ foo: 20 });
    }
    load();
  }, []);
  return (
    <MyContext.Provider value={state}>{props.children}</MyContext.Provider>
  );
};

function useExampleHook() {
  const state = React.useContext(MyContext);
  return { state };
}

describe("minimal example", () => {
  test("the example", async () => {
    const wrapper: WrapperComponent<{}> = ({ children }) => (
      <MyProvider>{children}</MyProvider>
    );
    const { result } = renderHook(() => useExampleHook(), {
      wrapper,
    });
    expect(result.current.state).toEqual({foo: 0});
  });
});
```
The test succeeds but emits a warning as described above.
Notice that `useEffect` calls an async function `load` which calls an async function `loadData`.
Note that `loadData` does *nothing* and, although marked as async, is not async.

Now the following happens:
- if I remove the call to `loadData()` there are no warnings.
- if I don't await `loadData` there are no warnings.
- if I return `new Promise(resolve => setTimeout(() => resolve(), 1))` from `loadData` there are no warnings.
- if I return `new Promise(resolve => resolve())` from `loadData` there *are* warnings.
- if I instead of calling `loadData` call `await new Promise(resolve => resolve())` then there *are* warnings.

It seems to be the presence of immediately-resolving promises in the `useEffect` callback that trips it up. I have no idea why.

This is very strange! I hope you have an idea of what is happening here :)

Now, the funny thing is that if I remove the call to `loadData` (which does *nothing*), then no warnings are emitted.
It seems I can do normal async/await inside `load`, but I cannot call any other async functions.

**Comment by adamschoenemann** [comment]
I copied my test to a fresh dir with create-react-app and it no longer has the weird behaviour... I think there might be some packages interefering with each other on my end. Sorry to waste your time

**Comment by adamschoenemann** [comment]
Welp, sorry! After banging my head at the wall for so long I seem to have lost my wits. The copied test did *indeed* warn on a fresh create-react-app, I was just too stupid to notice.

I've made a minimal example here: https://github.com/adamschoenemann/react-hooks-testing-library-issue-809. 
Just run `npm install && npm test` and you should see the test pass but with a big `console.error`

**Comment by adamschoenemann** [comment]
Any progress on this? Could you reproduce using the minimal example I provided?

**Comment by mpeyper** [comment]
Sorry for the delay, I looked when you first put it up, but didn’t have time to write a response and it went on the back burner.

Long story short, the issue is the `await` in the provider’s `load` effect ([here](https://github.com/adamschoenemann/react-hooks-testing-library-issue-809/blob/fbf79a85e0972944f01511924801692d4df8fd8d/src/App.test.tsx#L11) delaying the ‘setState` until after the test has finished. Even though the `loadData` function is not `async` itself, it still gets queued because it’s called with `await`. 

Basically, the provider is rendered like any other component so any behaviour it has will affect the test output. It’s a bit of an awkward situation because it’s not what your testing, but will need put [a wait](https://react-hooks-testing-library.com/reference/api#async-utilities) into your test after the `expect` to prevent the warning from react.

**Comment by aplassen** [comment]
Same issue here. Can confirm that the `await waitForNextUpdate()` at the end of the test block made the error go away.  Thanks for the workaround, @mpeyper 👍
